src/librustc_lint/builtin.rs

```diff
@@ -1544,7 +1544,8 @@ fn check_const(cx: &LateContext, body_id: hir::BodyId, what: &str) {
                 span,
                 &format!("this {} likely exhibits undefined behavior. \
                 The rules on what exactly is undefined behavior aren't clear, \
-                so this check might be ovezealous.", what),
+                so this check might be ovezealous. Please open an issue on the rust compiler
+                repository if you believe it should not be considered undefined behavior", what),
```

Comment:
Missing trailing `.`

Reply:
Also, so far all the UB we detect is about violating the layout, right? Should it say that?

Reply:
I don't really want to expose "data layout invariants" to users

Reply:
Users writing code on this level *have* to think about stuff like that. I am not sure what the best terminology is, but we are already exposing these invariants to users writing unsafe code.

Reply:
It's also erroring if a character is not a character, and I want to do more things like vtables not being vtables, lengths not being integers, ... so it's not just layout stuff

Reply:
> It's also erroring if a character is not a character

That's a layout invariant (`char` not being the same as `u32`, just like `bool` isn't the same as `u8`).

> more things like vtables not being vtables, lengths not being integers, ... so it's not just layout stuff

The latter is also a layout invariant. vtables are a bit more tricky but I think can also be put into that category.

Maybe the term isn't great. I also sometimes call these "bit-level invariants" because they only concern the shape of the bits and are entirely local; they don't say anything about ownership or aliasing or the current state of memory (i.e., which addresses are in-bounds of an allocation).

Reply:
> which addresses are in-bounds of an allocation

We are in fact checking this for references ;)

> aliasing

We could totally check whether there are multiple `&mut` pointing to the same memory, even across statics.

> The latter is also a layout invariant.

slice lengths are just `usize`, and in general that can be a pointer. So we'd be looking at the type and thus enforce the second field to be an actual integer.